Add support for Tornado >= 5

andrewrabert opened the pull request:
Also included are a number of beneficial changes I discovered need for while working on this.

andrewrabert committed:
Add support for Tornado >= 5

andrewrabert committed:
Pin Tornado based on Python version

noone234 commented on sprockets/http/app.py:
```diff
@@ -40,13 +45,18 @@ def on_shutdown_ready(self):
 
+    def _ioloop_has_tasks(self):
+        if ASYNCIO_TORNADO:
+            return bool(asyncio.Task.all_tasks())
```
Question: Can there be an async task outside of Tornado's ioloop? Is it possible? Likely?

If it is impossible, then this is great. If it is possible but not likely, then this is good for a first cut, and it can always improve later.

andrewrabert replied:
Yes - it's possible. Likely for our use? No.

noone234 commented on sprockets/http/app.py:
```diff
@@ -120,6 +130,10 @@ def stop(self, io_loop):
+
+                if ASYNCIO_TORNADO and asyncio.iscoroutine(maybe_future):
+                        maybe_future = asyncio.create_task(maybe_future)
+
```
Sorry, I am not well versed in Tornado's concurrent module. In an effort to learn, I have a question. In the new conditional block, would it make sense to set running_async = True? Because we know that maybe_future is a coroutine, does it follow that it is also "future"?

noone234 commented on sprockets/http/app.py:
```diff
@@ -120,6 +130,10 @@ def stop(self, io_loop):
+                if ASYNCIO_TORNADO and asyncio.iscoroutine(maybe_future):
+                        maybe_future = asyncio.create_task(maybe_future)
+
                 if concurrent.is_future(maybe_future):
```
Sorry, I am not well versed in Tornado's concurrent module. Please tell me if I understand this right.

Following this code...

- Suppose that ASYNCIO_TORNADO is True, and maybe_future was a coroutine on line 134.
- In Tornado's "concurrent.py",
- - FUTURES includes concurrent.future.Futures and asyncio.Future
- - The is_future() method checks if its argument is an instance of FUTURES.
- Therefore if maybe_future is an instance of asyncio.Future, the condition on line 137 here will evaluate to true, and we will enter the code on line 139 that sets running_async to True.

noone234 commented on tests.py:
```diff
@@ -448,6 +448,7 @@ def test_that_signal_handler_invokes_shutdown(self):
             self.io_loop.add_callback_from_signal.assert_called_once_with(
                 runner._shutdown)
 
+    @unittest.skipUnless(tornado.version_info < (5,), '')
```
This makes sense. There are no callbacks in asyncio, right?

andrewrabert committed:
Reorganize and update requirements
- Move test deps to testing.txt
- Move doc deps to docs.txt. IMO - keeps things clearer
- Unpin flake8
- Require sprockets.http installation in edit mode

andrewrabert committed:
Drop support for Python < 3.5

andrewrabert committed:
Drop pypy support

andrewrabert committed:
Support only tornado>=5.0
- Drop support for all versions of tornado prior to 5.0. This negates
the need for version-aware conditionals that previous existed.
- Update tox to test Tornado 5.0 and the unpinned latest version.

andrewrabert committed:
Remove mock and update tests

andrewrabert committed:
Remove explicit object inheritance

andrewrabert committed:
Use python 3 super

andrewrabert committed:
Simplify setup.py

andrewrabert committed:
Ensure docs build in CI

andrewrabert committed:
Add missing changelog link

andrewrabert committed:
Ensure flake8 compliance in CI

andrewrabert committed:
Simplify doc config

gmr commented:
So one issue with the updates here is it seems you've jumped straight to Tornado 5 only. I'm not sold that we should not use Tornado's abstractions instead of going straight to asyncio usage.

andrewrabert commented:
@gmr - There's no abstraction of equal functionality in Tornado>=5.  :(

As for raising the minimum to 5, I did so for the sake of simplicity. Supporting older versions doesn't seem worth having dual logic as I did in https://github.com/sprockets/sprockets.http/pull/29/commits/74aaf7bf013d74c3ea4c135b250feae9211cd32b

Do you wish for me to add support back for that?
